### PR TITLE
Fix build errors on MSVC

### PR DESCRIPTION
**Pull request by ghost**
This wraps some gnu specific things so that people can build on MSVC. I wanted to test this interpreter out, but had to make these changes first.

### CODE REVIEW

File: Python/ceval.c [hunk 1]
```diff
@@ -5012,8 +5024,12 @@ call_function_ceval(PyThreadState *tstate, PyObject ***pp_stack, Py_ssize_t opar
 #else
     PyObject **stack = (*pp_stack) - nargs - nkwargs;
 #endif
-
+#ifndef _MSC_VER
     if (__builtin_expect(tstate->use_tracing, 0)) {
+#else
+     if ((tstate->use_tracing, 0)) {
```
Comment (kmod) [inline Python/ceval.c]: I think you need to remove the `0` (and parenthesis) here. Also the whitespace seems a little bit funky

File: Include/ceval.h [hunk 2]
```diff
@@ -87,8 +87,8 @@ PyAPI_FUNC(void) Py_SetRecursionLimit(int);
 PyAPI_FUNC(int) Py_GetRecursionLimit(void);
 
 #if !PY_DEBUGGING_CHECKS
-#define Py_EnterRecursiveCall(where) (0)
-#define Py_LeaveRecursiveCall(where) ((void)0)
+#define Py_EnterRecursiveCall(...) (0)
+#define Py_LeaveRecursiveCall(..) ((void)0)
```
Comment (kmod) [inline Include/ceval.h]: I'm curious, why is this needed? Also are there supposed to be 3 dots and then 2 dots?

File: Objects/obmalloc.c [hunk 3]
```diff
@@ -145,7 +145,16 @@ _PyObject_ArenaVirtualFree(void *ctx, void *ptr, size_t size)
 {
     VirtualFree(ptr, 0, MEM_RELEASE);
 }
+static void *
```
Comment (kmod) [inline Objects/obmalloc.c:148]: I see what you're doing here but I think a better fix might be to change the callers of _PyObject_ArenaMmap to call the correct function

File: Python/sysmodule.c [hunk 4]
```diff
@@ -1732,9 +1732,13 @@ static Py_ssize_t
 sys_getallocatedblocks_impl(PyObject *module)
 /*[clinic end generated code: output=f0c4e873f0b6dcf7 input=dab13ee346a0673e]*/
 {
+    #if PY_DEBUGGING_FEATURES
```
Comment (kmod) [inline Python/sysmodule.c:1735]: I'm curious, what's the link error here?

**Review by kmod — CHANGES_REQUESTED** [review]
Thanks so much for looking into this! I left some comments on your changes.

I think the bigger comment is that, sadly, windows support will be much harder than just fixing the build errors, since our JIT will need to be updated and I'm not sure what else. But this seems like a good first step!

File: Include/ceval.h (hunk 2 above)
Comment (ghost) [reply Include/ceval.h]: Errors out on wrong number of arguments to the macro on MSVC.
The two dots is my mistake

File: Python/sysmodule.c (hunk 4 above)
Comment (ghost) [reply Python/sysmodule.c:1735]: sys_getallocatedblocks_impl gets referenced later despite being ifdefed out. I stubbed to remove the error instead of removing the reference since its a documented member of sys

File: Objects/obmalloc.c (hunk 3 above)
Comment (ghost) [reply Objects/obmalloc.c:148]: I think just having a single wrapper around virtual memory allocation/deallocation is more convenient and makes more sense overall.

**Comment by ghost** [comment]
I moved over just the changes I had that were related to getting it to compile to this repo, I seem to have made two mistakes doing so, those are now corrected.

File: Include/ceval.h (hunk 2 above)
Comment (kmod) [reply Include/ceval.h]: What I meant was "why is a difference from CPython needed here", and it looks like the answer is because I messed up the conversion of Py_LeaveRecursiveCall. I think the proper solution here is to keep the definition of Py_EnterRecursiveCall and to remove the erroneous argument from Py_LeaveRecursiveCall

File: Objects/obmalloc.c (hunk 3 above)
Comment (kmod) [reply Objects/obmalloc.c:148]: Sure it could be a function wrapper or a macro wrapper, but I think it's incorrect to say that Mmap is implemented using VirtualAlloc.

File: Python/sysmodule.c (hunk 4 above)
Comment (kmod) [reply Python/sysmodule.c:1735]: I'm confused about what's going on here, is MSVC trying to compile ifdef'd out code? What's the error you get?

I removed the function because I believe it's better to not have the function than to have it behave incorrectly. It looks like the removal was unnecessary, so if you want you could just remove the `PY_DEBUGGING_FEATURES` ifdefs.
